test/test_common.c

```diff
@@ -551,65 +553,103 @@ static unsigned char xorstr(const char *str) {
     return result;
 }
 
-/* sanitize, sort, and compare */
-#define MAX_LINE_LENGTH  4096
-#define MAX_LINES 4096
-void assert_json_equal(const char* filename1, const char* filename2, int ignoreListOrder) {
-    char line1[MAX_LINE_LENGTH];
-    char line2[MAX_LINE_LENGTH];
-    char* lines1[MAX_LINES];
-    char* lines2[MAX_LINES];
-    int lineCount1 = 0;
-    int lineCount2 = 0;
-    int res = -1;
-
-    FILE* file1 = fopen(filename1, "r");
-    ASSERT_TRUE(file1, "Failed to open file: %s", filename1);
-
-    FILE* file2 = fopen(filename2, "r");
-    ASSERT_TRUE(file2, "Failed to open file: %s", filename2);
-
-    while (fgets(line1, MAX_LINE_LENGTH, file1)) {
-        sanitizeString(line1, ",{}[] \t\n"); /* roughly sanitize */
-        if (strlen(line1) != 0)
-            lines1[lineCount1++] = strdup(line1);
+/*
+ * Start a python service that reads JSON filename from its STDIN, generates its
+ * signature, ignoring elements order, and prints SHA256 signature to its STDOUT.
+ * The reason it is running as a service is to avoid the overhead of starting a
+ * new process for each evaluation and the intensive use of it.
+ */
+#define BUFFER_SIZE 1024
+static int pipe_in[2], pipe_out[2];
+static pid_t pid = -1;
+void start_json_sign_service() {
+    if (pipe(pipe_in) == -1 || pipe(pipe_out) == -1) {
+        perror("pipe failed");
+        exit(1);
     }
 
-    while (fgets(line2, MAX_LINE_LENGTH, file2)) {
-        sanitizeString(line2, ",{}[] \t\n"); /* roughly sanitize */
-        if (strlen(line2) != 0)
-            lines2[lineCount2++] = strdup(line2);
+    if ( (pid = fork()) == -1) {
+        perror("fork failed");
+        exit(1);
     }
 
-    if (lineCount1 != lineCount2) goto end_cmp;
+    if (pid == 0) {
+        /*child*/
+        close(pipe_in[1]);  /* Close write end of pipe_in */
+        close(pipe_out[0]); /* Close read end of pipe_out */
+        dup2(pipe_in[0], STDIN_FILENO); /* Redirect child STDIN */
+        dup2(pipe_out[1], STDOUT_FILENO); /* Redirect child STDOUT */
+        /* now replace child process with python service */
+        execlp("python3", "python3", "./test/json_signature_generator.py", NULL);
+        perror("execlp");
+        exit(1);
+    } else {
+        /*Parent*/
+        close(pipe_in[0]);  /* Close read end of pipe_in */
+        close(pipe_out[1]); /* Close write end of pipe_out */
+    }
```

Comment:
Not sure about the logic here but please consider if there is a potential race condition here. e.g. parent exists and starts running tests while child is still in the above "if check"

Reply:
The parent will block as long as it didn't read a single byte. I tested it by putting 10 seconds delay at the begning of the python code and the parent waited peacefully.